BoxContentSDK/BoxContentSDK/Cache/BOXRequestCache.m

```diff
@@ -0,0 +1,87 @@
+//
+//  BOXRequestCache.m
+//  BoxContentSDK
+//
+//  Created on 8/25/15.
+//  Copyright (c) 2015 Box. All rights reserved.
+//
+
+#import "BOXRequestCache.h"
+#import "BOXContentClient.h"
+#import "BOXUser.h"
+#import "BOXRequest_Private.h"
+#import "BOXFolderItemsRequest.h"
+
+#define CACHE_FOLDER @"boxRequestCache"
+
+@interface BOXRequestCache ()
+
+@property (nonatomic, readwrite, strong) NSString *cachePath;
+@property (nonatomic, readwrite, strong) PINCache *cache;
+
+@end
+
+@implementation BOXRequestCache
+
+- (instancetype) initWithUserID:(NSString *)userID
+{
+    return [self initWithUserID:userID cacheDirectory:nil];
+}
+
+- (instancetype) initWithUserID:(NSString *)userID cacheDirectory:(NSURL *)cacheDirectory
+{
+    self = [super init];
+    
+    if (self) {
+        NSFileManager *fileManager = [NSFileManager defaultManager];
+        
+        if (!cacheDirectory) {
+            cacheDirectory = [NSURL URLWithString:[NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES) objectAtIndex:0]];
+        }
+        
+        _cachePath = [[cacheDirectory path] stringByAppendingPathComponent:[NSString stringWithFormat:@"%@/%@", CACHE_FOLDER, userID]];
+        
+        if (![fileManager fileExistsAtPath:_cachePath]) {
+            [fileManager createDirectoryAtPath:_cachePath
+                   withIntermediateDirectories:YES
+                                    attributes:nil
+                                         error:nil];
+        }
+        _cache = [[PINCache alloc] initWithName:@"BoxRequestCache" rootPath:_cachePath];
+    }
+    
+    return self;
+}
+
+- (void)fetchCacheResponseForRequest:(BOXRequest *)request cacheBlock:(void(^)(NSDictionary *dictionary))cacheBlock
+{
+    PINCacheObjectBlock localCacheBlock = ^void(PINCache *cache, NSString *key, id __nullable object) {
+        if ([object isKindOfClass:[NSDictionary class]]) {
+            cacheBlock(object);
+        }
+    };
+    [self.cache objectForKey:[self keyForRequest:request] block:localCacheBlock];
+}
+
+- (void)removeCacheResponseForRequest:(BOXRequest *)request
+{
+    [self.cache removeObjectForKey:[self keyForRequest:request]];
+}
+
+- (void)updateCacheForRequest:(BOXRequest *)request withResponse:(NSDictionary *)JSONDictionary
+{
+    [self.cache setObject:JSONDictionary forKey:[self keyForRequest:request] block:nil];
```

Comment:
Should we check if ```keyForRequest``` is non-nil? if PINCache's ```setObject:forKey:``` is nil safe then we should be OK.